### PR TITLE
Map difference

### PR DESCRIPTION
**Pull request by dlesbre**
Add two difference functions:
- `domain_difference m1 m2` which is just `m1` restricted to elements that don't appear in `m2`. On set, this is the `Stdlib`'s `diff` function. Also defined for `WithForeign` operations.
- `difference f m1 m2` which returns the set of bindings that differ in `m1` and `m2`, combining bindings present in both with `f`.

### CODE REVIEW

File: src/signatures.ml [hunk 1]
```diff
@@ -424,10 +424,34 @@ module type BASE_MAP = sig
       but [f.f] can return [None] to remove a binding from the resutling map. *)
 
   type ('map1, 'map2, 'map3) polymerge = {
-    f : 'a. 'a key -> ('a, 'map1) value option -> ('a, 'map2) value option -> ('a, 'map3) value  option; } [@@unboxed]
+    f : 'a. 'a key -> ('a, 'map1) value option -> ('a, 'map2) value option -> ('a, 'map3) value option; } [@@unboxed]
   val slow_merge : ('map1, 'map2, 'map3) polymerge -> 'map1 t -> 'map2 t -> 'map3 t
   (** This is the same as {{: https://ocaml.org/api/Map.S.html#VALmerge}Stdlib.Map.S.merge} *)
 
+  val difference: ('a, 'a, 'a) polyinterfilter -> 'a t -> 'a t -> 'a t
```
Comment (mlemerre) [inline src/signatures.ml]: Maybe we should have type 'a difference = ('a,'a,'a) polyinterfilter?

File: src/signatures.ml [hunk 2]
```diff
@@ -424,10 +424,34 @@ module type BASE_MAP = sig
...
+  (** [difference f map1 map2] returns a map comprising of the bindings
```
Comment (mlemerre) [inline src/signatures.ml]: Maybe name this symmetric_difference? If I understood well what this function does

File: src/signatures.ml [hunk 3]
```diff
@@ -507,6 +531,13 @@ module type HETEROGENEOUS_MAP = sig
         {!update_multiple_from_foreign}, except that instead of updating for all
         keys in [m_from], it only updates for keys that are both in [m_from] and
         [m_to]. *)
+
+    val domain_difference: 'a t -> 'b Map2.t -> 'a t
```
Comment (mlemerre) [inline src/signatures.ml]: And this would be just difference

File: src/functors.ml [hunk 4]
```diff
@@ -612,6 +612,66 @@ module MakeCustomHeterogeneousMap
...
+        else if branches_before pb mb pa ma
+        then if (mb :> int) land (pa :> int) == 0
+          then branch ~prefix:pb ~branching_bit:mb ~tree0:(difference f ta tb0) ~tree1:tb1
+          else branch ~prefix:pb ~branching_bit:mb ~tree0:tb0 ~tree1:(difference f ta tb1)
+        else join (pa :> int) ta (pb :> int) tb
+
+  let rec domain_difference ta tb =
+      match NODE.view ta, NODE.view tb with
+      | Empty, _ | _, Empty -> ta
+      | Leaf{key;_},_ -> if mem key tb then empty else ta
```
Comment (mlemerre) [inline src/functors.ml]: Should we use a function on the values to know if we want to remove the binding or not?

File: src/signatures.ml (hunk 2 above)
Comment (dlesbre) [reply src/signatures.ml]: Oh yes that's not a bad idea

**Pull request merged**